### PR TITLE
`MetaTensor`: collate; decollate; dataset; dataloader; out=; indexing and iterating across batches

### PR DESCRIPTION
**Pull request by rijobro**
Add collation and decollation for `MetaTensor`. Also add support for `out=` used as a kwarg, e.g., `torch.add(a,b,out=c)`.

Non-breaking, so going into `dev`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.

### CODE REVIEW

File: monai/data/utils.py [hunk 1]
```diff
@@ -346,9 +347,15 @@ def list_data_collate(batch: Sequence):
             ret = {}
             for k in elem:
                 key = k
-                ret[key] = default_collate([d[key] for d in data])
-            return ret
-        return default_collate(data)
+                data_for_batch = [d[key] for d in data]
+                ret[key] = default_collate(data_for_batch)
+                if isinstance(ret[key], MetaTensor) and all(isinstance(d, MetaTensor) for d in data_for_batch):
```
Comment (Nic-Ma) [inline monai/data/utils.py:352]: May I know what's the `meta` in the collated `ret[key]` if not all the items are `MetaTensor`?

Thanks in advance.

**Review by Nic-Ma — APPROVED** [review]
Hi @rijobro ,

Thanks for the quick update.
I am wondering whether it's possible to give a **warning** if user doesn't use our `list_data_collate`? Otherwise, I feel it is easily ignored and lead to unknown errors..

Thanks.

File: monai/data/utils.py (hunk 1 above)
Comment (rijobro) [reply monai/data/utils.py:352]: I envisage it being the case where all of them or none of them will be `MetaTensor`.

**Comment by rijobro** [comment]
@ericspod @wyli @Nic-Ma I upgraded the logic so that a `MetaTensor` knows if it is a batch of data or not. This is so that indexing and iterating with batches of data return a subset of the metadata, as expected:

```python
dl = DataLoader(ds)
batch = next(iter(dl))
batch[0]  # should only return the 0th image and 0th metadata
batch[:, -1]  # should return all metadata
batch[..., -1]  # should return all metadata
next(iter(batch))  # should return 0th image and 0th metadata.

```

**Comment by rijobro** [comment]
@Nic-Ma could you run `/build`?

**Comment by Nic-Ma** [comment]
> 

Hi @rijobro ,

Thanks for your update.
I am wondering whether it's possible to give a warning if user doesn't use our list_data_collate?

Thanks.

**Comment by Nic-Ma** [comment]
> Hi @rijobro ,
> 
> Thanks for the quick update. I am wondering whether it's possible to give a **warning** if user doesn't use our `list_data_collate`? Otherwise, I feel it is easily ignored and lead to unknown errors..
> 
> Thanks.

Hi @rijobro ,

Thanks for your update.
I am wondering whether it's possible to give a warning if user doesn't use our list_data_collate?

Thanks.

**Comment by rijobro** [comment]
From the following code snippet, the first is good (uses our collation fn). The middle case can have a warning. But how would we do the last case? Or are you just hoping for a warning in the middle case?

```python
import torch
from monai.data.meta_tensor import MetaTensor
from monai.data.dataset import Dataset
from monai.data.dataloader import DataLoader
from torch.utils.data import DataLoader as TorchDataLoader

data = [MetaTensor(torch.rand((1, 20, 20))) for _ in range(5)]
ds = Dataset(data)

# good
dl = DataLoader(ds)
next(iter(dl))

# can report warning
dl_no_collate = DataLoader(ds, collate_fn=lambda x: x)
next(iter(dl_no_collate))

# can't report warning
torch_dl = TorchDataLoader(ds)
next(iter(torch_dl))
```

**Comment by Nic-Ma** [comment]
Hi @rijobro ,

Yeah, I was asking about the last case, I understand it's hard to give a warning there. Just worried about user experience.
If you don't have any ideas, I am OK to merge this PR first.

Thanks.

**Comment by rijobro** [comment]
I don't have any better ideas at the moment, but will certainly reflect on it. The `MetaTensor` docstring contains this info:

> When creating a batch with this class, use `monai.data.DataLoader` as opposed to `torch.utils.data.DataLoader`, as this will take care of collating the metadata properly.

**Pull request merged**